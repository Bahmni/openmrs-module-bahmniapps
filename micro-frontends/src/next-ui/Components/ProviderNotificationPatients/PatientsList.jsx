import React from "react";
import PatientListTitle from "./PatientListTitle";
import { Accordion, AccordionItem } from "carbon-components-react";
import PropTypes from "prop-types";
import "./PatientsList.scss";
import PatientListContent from "./PatientListContent";
import { calculateAgeFromEpochDOB } from "../../utils/utils";

const PatientsList = ({ patientListWithMedications, handleOnClick }) => {

  const renderPatientListTitle = (patient) => (
    <PatientListTitle
      noOfDrugs={patient.length}
      name={patient[0].name}
      age={calculateAgeFromEpochDOB(patient[0].date_of_birth)}
      gender={patient[0].gender}
      identifier={patient[0].identifier}
      patientUuid={patient[0].patient_uuid}
      visitUuid={patient[0].visit_uuid}
    />
  );

  const renderPatientListContent = (patient) => (
    patient.map((medication, index) => (
      <PatientListContent
        key={index}
        patientMedicationDetails={medication}
        handleOnClick={handleOnClick}
      />
    ))
  );

  return (
    <Accordion className="patient-list-dropdown">
      {patientListWithMedications && patientListWithMedications.map((patient, index) => (
        <AccordionItem key={index} title={renderPatientListTitle(patient)}>
          {renderPatientListContent(patient)}
        </AccordionItem>
      ))}
    </Accordion>
  );
};

PatientsList.propTypes = {
  patientListWithMedications: PropTypes.arrayOf(
    PropTypes.arrayOf(
      PropTypes.shape({
        administered_date_time: PropTypes.array.isRequired,
        administered_drug_name: PropTypes.string.isRequired,
        administered_dose: PropTypes.string.isRequired,
        administered_dose_units: PropTypes.string.isRequired,
        administered_route: PropTypes.string.isRequired,
        date_of_birth: PropTypes.array.isRequired,
        gender: PropTypes.string.isRequired,
        identifier: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        patient_uuid: PropTypes.string.isRequired,
        visit_uuid: PropTypes.string.isRequired,
        medication_administration_performer_uuid: PropTypes.string.isRequired,
        medication_administration_uuid: PropTypes.string.isRequired
      }).isRequired
    ).isRequired
  ).isRequired,
  handleOnClick: PropTypes.func.isRequired
};

export default PatientsList;