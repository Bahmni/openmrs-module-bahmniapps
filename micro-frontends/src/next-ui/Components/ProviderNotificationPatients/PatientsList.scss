.patient-list-dropdown {
    li.bx--accordion__item {
        padding: 0;
        margin-bottom: 12px;
        background-color: transparent;
        border: 1px solid #007d79;
    }

    li.bx--accordion__item:hover {
        border: transparent;
    }
}

.patient-list-content .content-info .notes {
    background-color: transparent;
}

.patient-list-content {
    border-bottom: 1px solid #C6C6C6;
}