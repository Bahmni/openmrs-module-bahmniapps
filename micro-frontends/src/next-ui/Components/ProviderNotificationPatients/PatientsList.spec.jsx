import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import PatientsList from './PatientsList';

describe('PatientsList Component', () => {
  const patientListWithMedications = [
    [{
      administered_date_time: '2022-06-09',
      administered_drug_name: 'Drug A',
      administered_dose: '10mg',
      administered_dose_units: 'mg',
      administered_route: 'Oral',
      date_of_birth: '1990-01-01',
      gender: 'M',
      identifier: 'PID123',
      name: 'John Doe',
      patient_uuid: 'patient123',
      visit_uuid: 'visit123',
      medication_administration_performer_uuid: 'performer123',
      medication_administration_uuid: 'medication123'
    }]
  ];

  const handleOnClickMock = jest.fn();

  beforeAll(() => {
    jest.useFakeTimers().setSystemTime(new Date('2024-06-09'));
  });

  it('renders the component without crashing', () => {
    const {container} = render(
      <PatientsList
        patientListWithMedications={patientListWithMedications}
        handleOnClick={handleOnClickMock}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders AccordionItem components for each patient', () => {
    const { getAllByRole } = render(
      <PatientsList
        patientListWithMedications={patientListWithMedications}
        handleOnClick={handleOnClickMock}
      />
    );
    const accordionItems = getAllByRole('button', { name: /John Doe - Male, 34 years 5 months 8 days/i });
    expect(accordionItems).toHaveLength(patientListWithMedications.length);
  });

  it('renders PatientListTitle with correct props', () => {
    const { getByText } = render(
      <PatientsList
        patientListWithMedications={patientListWithMedications}
        handleOnClick={handleOnClickMock}
      />
    );
    const patientName = getByText('John Doe - Male, 34 years 5 months 8 days');
    expect(patientName).toBeTruthy();
  });

  it('renders PatientListContent with correct props', () => {
    const { getByText } = render(
      <PatientsList
        patientListWithMedications={patientListWithMedications}
        handleOnClick={handleOnClickMock}
      />
    );
    const drugName = getByText('Drug A');
    expect(drugName).toBeTruthy();
  });
});
