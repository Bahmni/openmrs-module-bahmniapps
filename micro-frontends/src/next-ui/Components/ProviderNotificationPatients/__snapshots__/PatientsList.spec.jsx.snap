// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatientsList should render without crashing 1`] = `
<div>
  <ul
    class="bx--accordion patient-list-dropdown bx--accordion--end bx--accordion--md"
  />
</div>
`;
